pkg/operator/ceph/disruption/machinelabel/reconcile.go

```diff
@@ -0,0 +1,127 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machinelabel
+
+import (
+	"context"
+	"time"
+
+	"github.com/coreos/pkg/capnslog"
+	mapiv1 "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
+	"github.com/rook/rook/pkg/operator/ceph/disruption/controllerconfig"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	controllerName                  = "machinelabel-controller"
+	MachineFencingLabelKey          = "fencegroup.rook.io/cluster"
+	MachineFencingNamespaceLabelKey = "fencegroup.rook.io/clusterNamespace"
+	OSDPodlabelKey                  = "app"
+	OSDPodCephClusterLabelKey       = "rook_cluster"
+	OSDPodLabelValue                = "rook-ceph-osd"
+)
+
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", controllerName)
+
+type ReconcileMachineLabel struct {
+	scheme  *runtime.Scheme
+	client  client.Client
+	log     *capnslog.PackageLogger
+	options *controllerconfig.Context
+}
+
+type machine struct {
+	isOccupiedByOSD bool
+	RawMachine      mapiv1.Machine
+}
+
+// Reconcile is the implementation of reconcile function for ReconcileMachineLabel
+// which ensures that the machineLabel for the osd pods are in correct state
+func (r *ReconcileMachineLabel) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	r.log.Infof("reconciling %s", request.NamespacedName)
+
+	// Fetch list of osd pods for the requested ceph cluster
+	pods := &corev1.PodList{}
+	err := r.client.List(context.TODO(), pods, client.InNamespace(request.Namespace), client.MatchingLabels{OSDPodlabelKey: OSDPodLabelValue, OSDPodCephClusterLabelKey: request.Name})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	// Fetch list of machines available
+	machines := &mapiv1.MachineList{}
+	err = r.client.List(context.TODO(), machines, client.InNamespace("openshift-machine-api"))
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	nodeMachineMap := map[string]machine{}
+
+	r.log.Infof("Number of osd pods found %v", len(pods.Items))
+	r.log.Infof("Number of machines found %v", len(machines.Items))
+
+	// Adding machines to nodeMachineMap
+	for _, m := range machines.Items {
+		if m.Status.NodeRef != nil {
+			nodeMachineMap[m.Status.NodeRef.Name] = machine{RawMachine: m}
+		}
+	}
+
+	// Marking out machines that are occupied by the osd pods
+	for _, pod := range pods.Items {
+		if pod.Spec.NodeName != "" {
+			if val, p := nodeMachineMap[pod.Spec.NodeName]; p {
+				val.isOccupiedByOSD = true
```

Comment:
I'd suggest this code block to simplify:
```
if _, ok := nodeMachineMap[pod.Spec.NodeName]; ok {
  nodeMachineMap[pod.Spec.NodeName].isOccupiedByOSD = true
}
```

Reply:
was getting this error when i tried that `cannot assign to struct field nodeMachineMap[pod.Spec.NodeName].isOccupiedByOSD in mapgo`